phoenix-core/src/main/java/org/apache/phoenix/index/IndexMaintainer.java

```diff
@@ -883,6 +903,14 @@ public void readFields(DataInput input) throws IOException {
         dataEmptyKeyValueCF = Bytes.readByteArray(input);
         emptyKeyValueCFPtr = new ImmutableBytesPtr(Bytes.readByteArray(input));
         
+        int numIndexedExpressions = WritableUtils.readVInt(input);
```

Comment:
Also, the removal of the indexedColumnsTypes will be a problem. You'll need to not do that plus do your new serialization based on the hacky way I mentioned (we should protobuf this at the next major release to give us more flexibility here).